TEP-0092: TaskRun Timeouts - Updated Proposal (#948)
tektoncd/community: teps/0092-scheduling-timeout.md

---
status: implementable
title: TaskRun Timeouts
creation-date: '2021-10-13'
last-updated: '2023-02-06'
authors:
- '@bobcatfish'
- '@badamowicz' # author of the original issue (https://github.com/tektoncd/pipeline/issues/4078)
- '@emmamunley'

> - '@jerop'

Reviewer: could you add me as a tep collaborator? (we added a new "collaborators" field in addition to authors)

- [References](#references)
<!-- /toc -->

## Summary

Add runtime configuration options to allow setting specific timeouts for parts of a `TaskRun` dedicated to scheduling the pod vs executing the pod.

* The `scheduling` timeout applies from the start time of `TaskRun` until the scheduled time of the `Pod`.
* The `execution` timeout applies from the scheduled time of the `Pod` until the completion time of `TaskRun`.
* The `total` timeout is the combination of `scheduling` and `execution` timeouts. It applies from the start time of `TaskRun` until the completion time of `TaskRun`.

## Motivation

Current timeout logic doesn't take into account the time a `Pod` may spend waiting to be scheduled; the timeout countdown starts as soon as the `TaskRun` starts.

However, in some environments, the `Pod` backing the `TaskRun` may not start executing until
some time after the `TaskRun` has started.

### Goals

The goal of this proposal is to add configuration to `TaskRuns` which would allow for specifying:
* timeout for scheduling a `Pod` for a given `TaskRun`
* timeout for executing a `Pod` for a given `TaskRun`
* total time for scheduling and executing a given `TaskRun`

> <!-- To keep consistency among the start time vs creation time, both the `scheduling` and `total` timeout will start ticking at the startTime of the `TaskRun`, similar to `PipelineRuns`. In addition, to keep the consistency of timeout definitions

Reviewer: is this intended to be commented out?

Reviewer 2: it is meant to be removed @EmmaMunley

#### Queued Timeout
Using Tekton to get parity with a build system that provides a "pending" or "queued" timeout option, for example [Google Cloud Build's queueTtl][queuettl]. This could be viewed as subset of first use case: where the build system is running in a resource constrained environment.

## Requirements

* Allow the timeouts to be specified at runtime in `TaskRuns` and `PipelineRuns`.
* Allow the timeouts to differ between `TaskRuns`.

## Proposal

Add configuration to `TaskRuns` to allow configuring `scheduling`, `execution` and `overall` timeout of a `TaskRun`.

| Term                         | Definition                                               |
|:---------------------------- |:-------------------------------------------------------- |
| Start time of `TaskRun`      | `TaskRun.Status.StartTime`                               |
| Completion time of `TaskRun` | `TaskRun.Status.CompletionTime`                          |
| Scheduled time of `Pod`      | `Pod.Status.Conditions[PodScheduled].LastTransitionTime` |

### Scheduling Timeout

> The `scheduling` timeout of a `TaskRun` applies from the start time of `TaskRun` until the scheduled time of its `Pod`. The `execution` timeout has no default value.

Reviewer: should this read "the scheduling timeout has no default value"?

The [`PodScheduled`][pod-conditions] condition is set on a `Pod` indicating that a node is available for the `Pod` to run on and it has been assigned to that node.

### Execution Timeout

The `execution` timeout of a `TaskRun` applies from the scheduled time of its `Pod` until the completion time of `TaskRun`. The `execution` timeout has no default value.

### Total Timeout

The `total` timeout of a `TaskRun` is the combination of `scheduling` and `execution` timeouts. It applies from the start time of the `TaskRun` until the completion time of the `TaskRun`.

Defaults for `total` timeout:
- If either `execution` or `scheduling` timeout is provided, the `total` timeout will default to `60` mins

> - Otherwise, the default `total` timeout will be calculated from `scheduling` + `execution`.

Reviewer: Interesting, this is different from how pipelineruns work. Probably worth noting that, but I think this is a valid approach.

### API Changes

`TaskRun` has one timeout field ([docs][timeout-docs]):

```yaml
spec:
  timeout: 60s
```

To support the new timeouts we would follow, we would add a dictionary to hold configuration for `scheduling`, `execution`, and `total` timeout fields. The previous `timeout` field would be deprecated once this feature is stable.

```yaml
spec:
  timeouts:
    scheduling: "5m" # optional, if specified, from TaskRun.StartTime to Pod.ScheduledTime
    execution: "5m" # optional, if specified, from Pod.ScheduledTime to TaskRun.CompletionTime
    total: "10m" # optional, if specified, from TaskRun.StartTime to TaskRun.CompletionTime (defaults to 60m)
  timeout: "10m" # existing deprecated field, from TaskRun.StartTime to TaskRun.CompletionTime (defaults to 60m)
```

When set, the order of timeout application would be:
1. At the start of the `TaskRun`, the `scheduling` and `total` timeouts counters would begin
2. Once the `Pod` executing the `TaskRun` has been scheduled, the `scheduling` timeout would stop and the `execution` timeout would begin.
3. The completion of the `TaskRun` must be before the `execution` and `total` timeout counters end.

Note that:

> * If no timeouts are set, only the `total` timeout is used

Reviewer: ```suggestion
* If no timeouts are set, only the default `total` timeout is used
```

Reviewer: It looks like you might have accidentally marked this comment as resolved

* Note that the sum of `scheduling` timeout and `execution` timeout must equal the `total` timeout
* If either `scheduling` or `execution` timeout is set to 0, then total timeout must also be set to 0

### Examples

###### User has a resource-constrained environment and wants to cancel if not scheduled quickly

User creates this:

```yaml
spec:
  timeouts:
    scheduling: "5m"
```

The spec that is calculated is this:

```yaml
spec:
  timeouts:
    scheduling: "5m"  # User-specified value
    execution: "55m"  # Calculated from default total timeout - scheduling timeout
    total: "60m"  # Default value
```

###### User does not care when the `Pod` is scheduled but wants the `TaskRun` executed within a specific timeframe

User creates this:

```yaml
spec:
  timeouts:

>     execution: "20m"

Reviewer: Given the use case stated here, I think a user would be more likely to create this:
```
timeouts:
  scheduling: 0
  execution: 20m
  total: 0
```
if they truly don't care about how long it takes for a pod to be scheduled

spec:
  timeouts:
    scheduling: "5m"
    execution: "10"
    total: "20m"
```
Invalid: This is invalid because `scheduling` and `execution` timeout does not add up to `total` timeout.

```yaml
spec:
  timeouts:
    scheduling: "10m"
    execution: "75m"
    total: "0m"
```
Invalid: This is invalid because `scheduling` and `execution` timeout does not add up to `total` timeout.

### What if the pod gets rescheduled?

If the `Pod` gets rescheduled, the `scheduling` timer would not restart, i.e. it would still be considered already scheduled.

As a side note, we already do not handle this situation well, see [pipelines#2813](https://github.com/tektoncd/pipeline/issues/2813).

### PipelineRun

#### TaskRunTemplate

In [TEP-0119][tep-0119], we added `TaskRunTemplate` to support setting configuration for all `TaskRun` in `PipelineRun`.  This proposal adds a new timeout configuration within a `TaskRunTemplate` that allows users to specify `scheduling`, `execution` and `total` timeout for all `TaskRuns` executed in a given `PipelineRun`.

```yaml=
apiVersion: tekton.dev/v1beta1
kind: PipelineRun
metadata:
  generateName: clone-test-build-
spec:
  pipelineRef:
    name: clone-test-build

>   taskRunTemplate:

Reviewer: I'm curious what would happen if you used this TaskRun template in addition to pipeline task timeouts. This could create some confusing results. e.g. with the below template, using a pipeline task timeout of 5m would be invalid.

Reviewer 2: I think that runtime specifications should take precedence over authoring time specifications

taskrunspecs > taskruntemplate > pipelinetask 

but we'd have to be clear about it in the documentation

wonder what happens if retries are set in both taskrunspecs/taskruntemplate and pipelinetask 🤔 

also open to making this scenario validation error initially

#### Flexibility
* This proposal does not involve bringing in any new dependencies, coupling Pipelines to any new tools, or introducing templating logic
* Again implementing this very specific timeout configuration in our API does open the door for adding more of this configuration and the proposal is not very flexible in that it adapting it to new timeout cases will mean adding new config options and explaining in detail how they relate to the existing ones

#### Conformance

* Likely we would include this timeout configuration in our conformance surface for `PipelineRuns` and `TaskRuns`.
* If something other than Kubernetes was used to back the API, the implementors could make their own decisions about what `scheduling` means in that environment. The recommendation would be to stop the timer once the execution of the `TaskRun` is ready to begin.

#### Performance

* Additional timeouts to consider will mean additional `TaskRuns` queued for later reconciling within the controller.
* This would mean at most 2x the number of timeouts for each individual `TaskRun`.

## Alternatives

### Configuration Levels

* Configure these timeouts at the controller level as well

>   * Pro: easy to set and forget in an environment that is always resource constrained

Reviewer: It looks like this alternative is the same as the alternative "configure at the controller level instead", can their pros and cons be combined?

    - [TaskRunSpec](#taskrunspec)
    - [TaskRunTemplate and TaskRunSpec](#taskruntemplate-and-taskrunspec)
  - [Pipeline](#pipeline)
  - [Notes and Caveats](#notes-and-caveats)
    - [Pending State](#pending-state)
    - [Timeouts in Kubernetes](#timeouts-in-kubernetes)
  - [Risks and Mitigations](#risks-and-mitigations)
- [Design Evaluation](#design-evaluation)
    - [Reusability:](#reusability)
    - [Simplicity](#simplicity)
    - [Flexibility](#flexibility)
    - [Conformance](#conformance)
    - [Performance](#performance)
- [Alternatives](#alternatives)
  - [Configuration Levels](#configuration-levels)
  - [Scheduling Timeout Boundaries](#scheduling-timeout-boundaries)
  - [Continue timeout until &quot;running&quot;](#continue-timeout-until-running)
  - [Leverage pending status](#leverage-pending-status)
  - [Syntax Options](#syntax-options)
  - [PipelineRun](#pipelinerun-1)
- [Test Plan](#test-plan)
- [Upgrade &amp; Migration Strategy](#upgrade--migration-strategy)
- [References](#references)
<!-- /toc -->

## Summary

Add runtime configuration options to allow setting specific timeouts for parts of a `TaskRun` dedicated to scheduling the pod vs executing the pod.

> * The `scheduling` timeout applies from the start time of `TaskRun` until the scheduled time of the `Pod`.

Reviewer 3: Do we mean "start time" in the status, or creation time ? I am asking this because, let's assume the controller is slow to pick up things, do we want to apply the timeout based of the creation time (when the user created the object) or when the controller first picked it up ?

* timeout for scheduling a `Pod` for a given `TaskRun`
* timeout for executing a `Pod` for a given `TaskRun`
* total time for scheduling and executing a given `TaskRun`

### Non-Goals

* Support any timeout configuration around any other `Pod` or `Container` states
* Support any timeouts related to the queuing of any types other than `TaskRuns` (e.g. `PipelineRuns`)

### Use Cases

#### Running Many `Tasks` in a Resource Constrained Environment
1. Build the artifacts of a large Java application
2. Start a `Pipeline` that starts some 35 Tasks in parallel, each performing Selenium tests against the deployed application. Each of these `Tasks` consists of:
  * a `Pod` with Websphere Liberty server along with the deployed application
  * an Oracle `Pod` as `Sidecar`
  * a `Pod` with a mock server as `Sidecar`
3. Do some cleanup
   Those 35 parallel `TaskRuns` cause a huge peak of CPU and memory consumption on the cluster's nodes and may eventually force `Pods` of other `PipelineRuns` to go to status `Pending`. This is expected behavior and OK for us. However, we had to increase all timeouts to avoid those pending Pods being killed due to timeout.

#### Fire and Forget `Runs`

> Supporting an environment where users may "fire and forget" a number of `PipelineRuns` or `TaskRuns`, more than the environment can handle, causing the `Runs` to be queued such that eventually the user no longer cares about these older queued `Runs` and wants them to naturally expire to make way for newer, more relevant `Runs`

Reviewer 3: I have trouble linking this use case with this timeout's proposal 🤔

## Proposal

Add configuration to `TaskRuns` to allow configuring `scheduling`, `execution` and `overall` timeout of a `TaskRun`.

| Term                         | Definition                                               |
|:---------------------------- |:-------------------------------------------------------- |
| Start time of `TaskRun`      | `TaskRun.Status.StartTime`                               |
| Completion time of `TaskRun` | `TaskRun.Status.CompletionTime`                          |
| Scheduled time of `Pod`      | `Pod.Status.Conditions[PodScheduled].LastTransitionTime` |

### Scheduling Timeout

The `scheduling` timeout of a `TaskRun` applies from the start time of `TaskRun` until the scheduled time of its `Pod`. The `scheduling` timeout has no default value.

The [`PodScheduled`][pod-conditions] condition is set on a `Pod` indicating that a node is available for the `Pod` to run on and it has been assigned to that node.

### Execution Timeout

The `execution` timeout of a `TaskRun` applies from the scheduled time of its `Pod` until the completion time of `TaskRun`. The `execution` timeout has no default value.

### Total Timeout

> The `total` timeout of a `TaskRun` is the combination of `scheduling` and `execution` timeouts. It applies from the start time of the `TaskRun` until the completion time of the `TaskRun`.

Reviewer 2: ```suggestion
The `total` timeout of a `TaskRun` is the combination of `scheduling` and `execution` timeouts. It applies from the start time of the `TaskRun` until the completion time of the `TaskRun`. This timeout is useful when a user doesn't care to distinguish between scheduling vs execution, they only want to general timeout of a `TaskRun`
```

Defaults for `total` timeout:

> - If either `execution` or `scheduling` timeout is provided, the `total` timeout will default to `60` mins

Reviewer 3: Ideally, it would default to the default value configured in the default configmap 👼🏼. I think that's what implied but we may want to be explicit.

- Otherwise, the default `total` timeout will be calculated from `scheduling` + `execution`.

> * Note that this is different from how `PipelineRuns` works

Reviewer: ```suggestion
* Note that this is different from how `pipelineRun.spec.timeouts` works: It's possible to declare PipelineRun timeouts where `timeouts.tasks` + `timeouts.finally` < `timeouts.pipeline`, and `timeouts.pipeline` always defaults to 60m and is never calculated by adding `timeouts.tasks` and `timeouts.finally`.
```

> * Note that this is different from how `pipelineRun.spec.timeouts` works: It's possible to declare PipelineRun timeouts where `timeouts.tasks` + `timeouts.finally` < `timeouts.pipeline`, and `timeouts.pipeline` always defaults to 60m and is never calculated by adding `timeouts.tasks` and `timeouts.finally`.

Reviewer 3: I am not sure we need that note here. It does confuse me more than anything, as we are talking about `Task` and `TaskRun` here, so talking about `PipelineRun` seems out of place 😅

### API Changes

`TaskRun` has one timeout field ([docs][timeout-docs]):

```yaml
spec:
  timeout: 60s
```

To support the new timeouts we would follow, we would add a dictionary to hold configuration for `scheduling`, `execution`, and `total` timeout fields. The previous `timeout` field would be deprecated once this feature is stable.

```yaml
spec:
  timeouts:
    scheduling: "5m" # optional, if specified, from TaskRun.StartTime to Pod.ScheduledTime
    execution: "5m" # optional, if specified, from Pod.ScheduledTime to TaskRun.CompletionTime
    total: "10m" # optional, if specified, from TaskRun.StartTime to TaskRun.CompletionTime (defaults to 60m)

>   timeout: "10m" # existing deprecated field, from TaskRun.StartTime to TaskRun.CompletionTime (defaults to 60m)

Reviewer 3: If `timeout` is different that `timeouts.total`, which one takes "priority" ? Or would it be disallowed to specify `timeout` if any `timeouts.*` is specified ?

Author: It would be disallowed to specify timeout if any timeouts.* is specified.

```

When set, the order of timeout application would be:
1. At the start of the `TaskRun`, the `scheduling` and `total` timeouts counters would begin
2. Once the `Pod` executing the `TaskRun` has been scheduled, the `scheduling` timeout would stop and the `execution` timeout would begin.
3. The completion of the `TaskRun` must be before the `execution` and `total` timeout counters end.

Note that:

* If no timeouts are set, only the default `total` timeout is used
* Note that the sum of `scheduling` timeout and `execution` timeout must equal the `total` timeout
* If either `scheduling` or `execution` timeout is set to 0, then total timeout must also be set to 0

### Examples

###### User has a resource-constrained environment and wants to cancel if not scheduled quickly

User creates this:

```yaml
spec:
  timeouts:
    scheduling: "5m"
```

The spec that is calculated is this:

```yaml
spec:
  timeouts:
    scheduling: "5m"  # User-specified value
    execution: "55m"  # Calculated from default total timeout - scheduling timeout

>     total: "60m"  # Default value

Reviewer 3: So this one is a bit weird/confusing to me. Ideally, I want the scheduling to not take more than `5m` and the whole *execution* of my Task (scheduling including) the default value, so 60m. My assumption is then that, the `TaskRun` wouldn't time-out before 60m has passed, no matter how much time the scheduling did take. This is, however, not what happens in the proposal. If the scheduling took about 10 second, then, the `TaskRun` would timeout after 56min and not 60min (because the execution is still gonna timeout after 55m, even though we had more time for the total).

Author: Thanks for the catch. Agreed. The execution time could range from 55 mins to 60 minutes depending on how long scheduling took. The TEP has been updated to reflect this.

```

###### User does not care when the `Pod` is scheduled but wants the `TaskRun` executed within a specific timeframe

User creates this:

```yaml
spec:
  timeouts:
    scheduling: "0m"
    execution: "20m"
    total: "0m"

```

The spec that is calculated is this:

```yaml
spec:
  timeouts:
    scheduling: "0m"  # User-specified value
    execution: "20m"  # User-specified value

>     total: "20m"   # Calculated from scheduling + execution timeout

Reviewer: ```suggestion
    total: "0"   # No total timeout
```

Reviewer 2: ah @lbernick already corrected it -- @EmmaMunley this needs to be fixed

>     total: "20m"   # Calculated from scheduling + execution timeout

Reviewer 2: this is confusing, I'd have expected the total to always remain "0m" which means it'd run until completion (no timeout) -- cc @EmmaMunley @lbernick

```yaml=
apiVersion: tekton.dev/v1beta1
kind: PipelineRun
metadata:
  generateName: clone-test-build-
spec:
  pipelineRef:
    name: clone-test-build
  taskRunTemplate:
    timeouts:
      scheduling: "10m"
      execution: "30m"
      total: "40m"
```

#### TaskRunSpec

In [TEP-0119][tep-0119], we added `TaskRunSpec` to support setting configuration for a specific `TaskRun` in `PipelineRun`. This proposal adds a new timeout configuration within a `TaskRunSpec` that allows users to specify a `scheduling`, `execution` and `total` timeout for a specific `TaskRun` using `TaskRunSpecs`.

```yaml=
apiVersion: tekton.dev/v1beta1
kind: PipelineRun
metadata:
  generateName: clone-test-build-
spec:
  pipelineRef:
    name: clone-test-build
  taskRunSpecs:
    - pipelineTaskName: clone
      timeouts:
        scheduling: "10m"
        execution: "30m"
        total: "40m"
```

> All the `TaskRuns` would have the specified timeouts applied.

Reviewer 3: In that example, only the `TaskRun` for the `clone` pipeline task would have these specified timeouts.

```yaml=
apiVersion: tekton.dev/v1beta1
kind: PipelineRun
metadata:
  generateName: clone-test-build-
spec:
  pipelineRef:
    name: clone-test-build
  taskRunTemplate:
    timeouts:
      scheduling: "10m"
      execution: "30m"
      total: "40m"
  taskRunSpecs:
    - pipelineTaskName: build
      timeouts:
        scheduling: "20m"
        execution: "40m"
        total: "60m"
```

The "clone" and "test" `TaskRuns` would have the timeouts in the `TaskRunTemplate`, while "build" `TaskRun` would have its timeout that's specified in the `TaskRunSpecs`.

### Pipeline

The existing `timeout` field in `Pipelines`  allows users to specify the timeout for a specific `TaskRun` executed from a `Pipeline`. That is, [`pipeline.pipelinetask.timeout`][pipeline-timeout] is passed to the `taskrun.timeout`.

```yaml=
apiVersion: tekton.dev/v1beta1
kind: Pipeline
metadata:
  name: clone-test-build
spec:
  tasks:
  - name: clone
    timeout: 10m
    taskRef:
      name: git-clone
  - name: test
    timeout: 5m
    runAfter:
    - clone
    taskRef:
      name: junit
```

> Given that `taskrun.timeouts.total` is replacing `taskrun.timeout`, the existing `pipeline.pipelinetask.timeout` will now be passed to the new `taskrun.timeouts.total`.

Reviewer 3: Do we plan to allow to specify scheduling and execution timeouts in there as well ?

## Design Evaluation

This is an evaluation of the above proposal against the [Tekton Design Principles][design-priciples].

#### Reusability:

Since this is being added as a runtime feature, `Pipelines` and `Tasks` will be just as reusable as they were before the feature. If we need to specify these new timeouts at the level of `Tasks` within a `Pipeline`, this would make the `Pipelines` slightly less reusable since we would be potentially mixing environment specific configuration into the `Pipeline` definition.

#### Simplicity
* This does make the timeout logic more complex, and as mentioned in [risks and mitigations](#risks-and-mitigations) this continues the trend of adding more and more complex timeout logic. One upside is that this is only in the runtime configuration and not (yet) the Pipeline or Task authoring time.
* This follows the same trend as the recent addition of [finally timeout configuration][finally-timeout]
* As we add more timeout options it may become more difficult to reason about which timeout is applying when

> * It is debatable that this feature is absolutely necessary; there are use cases to support it however it could potentially not be widely used and would cost additional cognitive overhead for all users trying to reason through the timeout logic

Reviewer 3: I definitely agree with that sentence 😅

## Summary

Add runtime configuration options to allow setting a timeout for `TaskRuns` which applies before the `TaskRun's`
underlying pod is scheduled (Scheduling), and only start our existing timeout counters (Execution) after this timeout has passed, if specified.
Add a Total Timeout which will can be set if users don't care about the distribution between scheduling and execution timeouts.
## Motivation

Current timeout logic doesn't take into account the time a pod may spend waiting to be scheduled; the timeout countdown
starts as soon as the TaskRun starts. In some environments the pod backing the TaskRun may not start executing until

> some period of time after that. In order to keep conistent timeout definitions across pipelines, if the timeout is

Reviewer: nit: keeping timeout definitions consistent between CRDs is more of a goal of the proposal than a motivation for it. There's also two components to consistency: starting at start time vs creation time, and the meaning of 0 timeout.

### Goals

> The goal of this proposal is to add configuration to both and `TaskRuns` which would allow for a timeout

Reviewer 2: ```suggestion
The goal of this proposal is to add configuration to both `PipelineRuns` and `TaskRuns` which would allow for a timeout
```

Reviewer: I think this is still unaddressed :)

   (could be viewed as subset of first use case: where the build system is running in a resource constrained environment)

## Requirements
* Allow the timeout to be specified at runtime and to differ between TaskRuns
#### Boundaries of the timeouts
This new scheduling timeout would apply from the startTime of the `TaskRun` until
[the Pod condition `PodScheduled`](https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#pod-conditions) has
been added to the pod, indicating that a node is available for the pod to run on and it has been assigned to that node.
The new execution timeout would apply from the scheduledTime until the completion of the task.

> The new total timeout would apply from the startTime until the completion of the task.

Reviewer: ```suggestion
The total timeout would apply from the startTime until the completion of the task, like the existing TaskRun timeout.
```

## Requirements

> * Allow the timeout to be specified at runtime and to differ between TaskRuns

Reviewer: ```suggestion
* Allow the timeouts to be specified at runtime and to differ between TaskRuns
```

## Proposal

 add configuration to both `PipelineRuns` and `TaskRuns` to allow configuring this new timeout for `TaskRuns`. When

> specified, this timeout would apply before any other timeouts are started.

Reviewer: This sentence is no longer true; we're adding two new timeouts, and it's a bit confusing to say they apply before any other timeouts are started.

>   * Con: this means if the `TaskRun` is queued by the controller for processing, it would not timeout. Suggest including

Reviewer: This is under the section "Other options", but we are proposing doing this. Maybe you could reword the other option to be starting from the creation time?

This feature would be introduced [at the alpha stability level](https://github.com/tektoncd/pipeline/blob/main/docs/install.md#alpha-features).

Total Timeout Default:
  - If either `Execution` or `Scheduling` timeout is provided, the `Total` timeout will default to `60` mins
  - Otherwise, the default `Total` timeout will be calculated from `Scheduling` + `Execution`

> ### Examples

Reviewer: Nice! This section is great!

Use case: resource-constrained and wants to cancel if not scheduled quickly.
User creates this:

```yaml
spec:
  timeouts:
    scheduling: "5m"
```

The spec that is calculated is this:

```yaml
spec:
  timeouts:
    scheduling: "5m"  # User-specified value
    execution: "55m"  # Calculated from default total timeout - scheduling timeout
    total: "60m"  # Default value
```

Use case: user does not care when the pod is scheduled but wants the `TaskRun` executed within a specific timeframe.
User creates this:

```yaml
spec:
  timeouts:
    execution: "20m"
```

The spec that is calculated is this:

```yaml
spec:
  timeouts:
    scheduling: "40m"  # Calculated from default total timeout - execution timeout
    execution: "20m"  # User-specified value
    total: "60m"  # Default value
```

Use case: only cares about the overall timeout.
User creates this:

```yaml
spec:
  timeouts:
    total: "45m"
```

The spec that is calculated is this:

```yaml
spec:
  timeouts:
    scheduling: "0m"  # Defaults to 0

>     execution: "45m"  # Calculated by total - scheduling

Reviewer: Here, scheduling and execution timeout should be unset. If it takes 45 minutes to schedule the pod, the taskrun will time out.

    total: "45m"  # User-specified value
```

### Validation

These are valid yaml files:

```yaml
spec:
  timeouts:
    execution: "55m"
    total: "0m"
```
Valid: In this scenario, the pod will never timeout before the pod is scheduled, but will timeout if the runtime is not completed within 55 minutes.

```yaml
spec:
  timeouts:
    scheduling: "5m"
    total: "0m"
```
Valid: In this scenario, the pod will only timeout if the pod isn't scheduled within 5 minutes.

```yaml
spec:
  timeouts:
    scheduling: "10m"
    execution: "75m"

>     total: "0m"

Reviewer: I don't think this would be valid. Here, the total time would need to be 85 minutes.

```
Valid: In this scenario, the pod will timeout if not scheduled in 10m or runtime exceeds 75 minutes.

```yaml
spec:
  timeouts:
```

> Valid: It's valid to not provide any timeouts. In this scenario, the total timeout will default to 60 mins.

Reviewer: It might be clearer to put this under "examples", and say "if the user doesn't specify any timeouts, here's what would be used." I don't think this is valid yaml.

```yaml
spec:
  timeouts:

>     scheduling: "5m" # optional, if specified, from startTime to scheduledTime

Reviewer: I can't comment on the sections you didn't update, but can you flesh out the "design details" section a bit more on what happens if the various timeouts are set to 0? In addition, the "PipelineRun" section also needs to be updated to reflect the "execution" timeout in addition to the "scheduling" timeout, and the "The field in TaskRuns that represents the existing timeout" section needs to be updated to reflect the fact that the execution timeout is different from the existing timeout.

Lastly, on line 211, "The previous `timeout` field would be deprecated once this feature is beta." should read "stable" instead of "beta", since taskrun.timeout is now stable and we can't replace it with a beta feature.

Reviewer: Sorry for squishing a lot of suggestions into one comment; it looks like the PipelineRun section still needs to be updated though.

IMO, it makes sense to have a "scheduling" option in `pipelinerun.spec.timeouts.taskruns` (since how long it takes to schedule a taskrun doesn't depend on the contents of the taskrun), but probably doesn't make sense to add `pipelinerun.spec.timeouts.execution`, because that will differ between taskruns. Either way, it's worth explaining the reasoning in the PipelineRuns section.

    execution: "5m" # optional, if specified, from scheduledTime to completionTime

>     total: "10m" # this option would have the same functionality as the existing `timeout` field

Reviewer: It would be helpful to give a few examples of how these fields will be validated/calculated/defaulted, based on what the user sets. Maybe in a new "examples" section?

```

When set, the order of timeout application would be:
1. At the start of the `TaskRun`, the `scheduling` timeout counter would begin
2. Once the pod executing the `TaskRun` has been scheduled, the `scheduling` timeout would stop and the `execution`
   timeout would begin

> When the `scheduling` timeout is not set for a `TaskRun`, the `execution` timeout counter would begin from the `startTime` of the `TaskRun`.

Reviewer: This is no longer true; the execution timeout doesn't change meaning depending on whether the scheduling timeout is set.

> If neither the `scheduling` or `execution` timeout is specified, the total timeout would function as previously from the `startTime` of the `TaskRun`.### What if the controller is totally overloaded?

Reviewer: ```suggestion
The total timeout functions starts from the `startTime` of the `TaskRun`, like the existing TaskRun timeout.
```

> ### What if the controller is totally overloaded?

Reviewer: Since we're now starting the scheduling timeout at start time rather than creation time, I don't think this section is true anymore.

Cons:
* The situations a user might want to address here (i.e. image pull backoff, long image pulls, contrained resource
  environment) seem like they're all fairly distinct, and the values for timeouts for each of these (e.g. a few seconds
  only for image pull backoff, maybe 5 minutes for scheduling in a resource constrained environment) users would want to
  provide might be very different, so bundling them into one option doesn't seem like the best way to address them

> Introducing the scheduling timeout initially and only including the time before the pod has the condition `PodScheduled`

Reviewer: Since we've changed the proposal a bit, I'm not sure that it still leaves open options to add more timeouts later. The original idea was that the scheduling timeout would run before all other timeouts, but now we're introducing a set of timeouts that must all add up to the total. So it's not clear that we could still add a meaningful "imagepullbackoff" timeout later.

Reviewer: I think the section here about being able to add a new timeout later on can just be deleted.

Here is a list of some possible names for the `scheduling` timeout:

* `pending`
* `queued`
* `unscheduled`
* `pre`
* `before`
* `system`
* `overhead`
* `setup`
* `init` (confusing b/c of init containers)
* `scheduling`

### PipelineRun

A new section `taskruns` would be added to
[the existing timeout configuration in PipelineRun][timeout-docs] with one new option called `scheduling`:

```yaml
spec:
  timeouts:
    pipeline: "0h15m0s"
    tasks: "0h10m0s"
    finally: "0h5m0s"
    taskruns: # this section would indicate that this timeout should be applied to all pipeline tasks
      scheduling: "1m" # this is the new option
```

## Test Plan

* Create an end to end tests that ultimately creates an unschedulable pod and verify the `scheduling` timeout is enforced.
* Create an end to end test (or reconciler test if you fake the times?) that uses both the `scheduling` timeout and the `execution` timeout, and ultimately the execution times out (e.g. the `Task` just sleeps) - ensure the full time of both timeouts has been counted, e.g. not just the `execution` timeout.

## Upgrade & Migration Strategy

* This feature would be introduced [at the alpha stability level][alpha]

> * The existing `TaskRun` timeout option would be deprecated once this feature reaches `beta` stability

Reviewer: ```suggestion
* The existing `TaskRun` timeout option would be deprecated once this feature reaches `stable` stability
```

Reviewer: It looks like you might have accidentally resolved this comment as well
